core/src/main/scala/kafka/log/Log.scala

```diff
@@ -2254,7 +2256,21 @@ class Log(@volatile var dir: File,
    */
   private[log] def replaceSegments(newSegments: Seq[LogSegment], oldSegments: Seq[LogSegment], isRecoveredSwapFile: Boolean = false): Unit = {
     lock synchronized {
+      if (newSegments.size > 1 && newSegments.exists(_.size == 0))
+        throw new IllegalArgumentException(s"Expected only one cleaned segment should be created after cleaning, but we got $newSegments")
```

Comment:
replaceSegments() is called during log recovery too. So, we want to write the error message in a more general way. Sth like "There should be only one new log segment if any of the new segment is empty, but we got $newSegments instead".